app/controllers/allies_controller.rb

```diff
@@ -19,65 +19,77 @@ def add
       user_id: current_user.id, ally_id: ally_id
     )
 
-    if allyship
-      allyship.update(status: User::ALLY_STATUS[:accepted])
+    pusher_type = allyship ? 'accepted_ally_request' : 'new_ally_request'
+
+    setup_allyship(allyship, ally_id)
+
+    remove_allyship_request(ally_id)
+
+    handle_allyship_notifications(pusher_type, ally_id)
+
+    respond_to do |format|
+      format.html { redirect_to :back }
+      format.json { head :no_content }
+    end
+  end
 
-      # Notify the user who made the request
-      pusher_type = 'accepted_ally_request'
+  def remove
+    user_id = current_user.id
+    ally_id = params[:ally_id].to_i
 
-      # Get rid of original new_ally_request notification
-      Notification.where(
-        userid: current_user.id,
-        uniqueid: "new_ally_request_#{ally_id}"
-      ).destroy_all
+    # Destroy allyship
+    Allyship.where(user_id: user_id, ally_id: ally_id).destroy_all
+
+    respond_to do |format|
+      format.html { redirect_to :back }
+      format.json { head :no_content }
+    end
+  end
+
+  private
+
+  def setup_allyship(allyship, ally_id)
+    if allyship
+      allyship.update(status: User::ALLY_STATUS[:accepted])
     else
       Allyship.create(
         user_id: current_user.id,
         ally_id: ally_id,
         status: User::ALLY_STATUS[:pending_from_ally]
       )
-
-      # Notify the user the request has been made to
-      pusher_type = 'new_ally_request'
     end
+  end
+
+  def remove_allyship_request(ally_id)
+    Notification.where(
+      userid: current_user.id,
+      uniqueid: "new_ally_request_#{ally_id}"
+    ).destroy_all
+  end
 
-    uniqueid = "#{pusher_type}_#{current_user.id}"
+  def handle_allyship_notifications(pusher_type, ally_id)
+    unique_id = "#{pusher_type}_#{current_user.id}"
     data = JSON.generate(
       user: current_user.name,
       userid: current_user.id,
       uid: current_user.uid,
       type: pusher_type,
-      uniqueid: uniqueid
+      uniqueid: unique_id
     )
 
+    send_allyship_notifications(unique_id, data, pusher_type, ally_id)
```

Comment:
Looks like you have to update this method to take four arguments! Right now it's only taking 3.

```
ArgumentError:
  wrong number of arguments (given 4, expected 3)
./app/controllers/allies_controller.rb:83:in `send_allyship_notifications'
./app/controllers/allies_controller.rb:80:in `handle_allyship_notifications'
./app/controllers/allies_controller.rb:28:in `add'
./app/controllers/application_controller.rb:31:in `block in with_timezone'
./app/controllers/application_controller.rb:31:in `with_timezone'
./spec/controllers/allies_controller_spec.rb:59:in `block (4 levels) in <top (required)>'
./spec/controllers/allies_controller_spec.rb:80:in `block (5 levels) in <top (required)>'
```

Reply:
Thanks for the review! Got this cleaned up and made the tests green again 😄